flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/InternalKeyContextImpl.java

```diff
@@ -72,6 +73,10 @@ public void setCurrentKey(@Nonnull K currentKey) {
 
     @Override
     public void setCurrentKeyGroupIndex(int currentKeyGroupIndex) {
+        if (!keyGroupRange.contains(currentKeyGroupIndex)) {
+            throw KeyGroupRangeOffsets.newIllegalKeyGroupException(
+                    currentKeyGroupIndex, keyGroupRange);
+        }
```

Comment:
I have run several tests in flink-benchmarks. The result are as follows:

- With this change:

```
Benchmark                    Mode  Cnt     Score     Error   Units
KeyByBenchmarks.arrayKeyBy  thrpt   30  4907.599 ± 412.976  ops/ms
KeyByBenchmarks.tupleKeyBy  thrpt   30  8157.488 ± 911.389  ops/ms

Benchmark                                  (stateBackend)   Mode  Cnt     Score    Error   Units
MemoryStateBackendBenchmark.stateBackends          MEMORY  thrpt  30  4749.925 ± 35.146  ops/ms
MemoryStateBackendBenchmark.stateBackends              FS  thrpt   30 4815.565 ± 27.069  ops/ms
MemoryStateBackendBenchmark.stateBackends        FS_ASYNC  thrpt  30  4741.209 ± 46.514  ops/ms

Benchmark                                 (stateBackend)   Mode  Cnt    Score   Error   Units
RocksStateBackendBenchmark.stateBackends           ROCKS  thrpt   30  289.627 ± 1.684  ops/ms
RocksStateBackendBenchmark.stateBackends       ROCKS_INC  thrpt   30  287.924 ± 4.893  ops/ms

```

- Without this change:

```
Benchmark                    Mode  Cnt     Score     Error   Units
KeyByBenchmarks.arrayKeyBy  thrpt   30  5004.647 ± 379.077  ops/ms
KeyByBenchmarks.tupleKeyBy  thrpt   30  8241.331 ± 870.276  ops/ms

Benchmark                                  (stateBackend)   Mode  Cnt     Score    Error   Units
MemoryStateBackendBenchmark.stateBackends          MEMORY  thrpt   30  4822.507 ± 65.687  ops/ms
MemoryStateBackendBenchmark.stateBackends              FS  thrpt   30  4748.527 ± 43.915  ops/ms
MemoryStateBackendBenchmark.stateBackends        FS_ASYNC  thrpt   30  4768.075 ± 41.401  ops/ms

Benchmark                                 (stateBackend)   Mode  Cnt    Score   Error   Units
RocksStateBackendBenchmark.stateBackends           ROCKS  thrpt   30  287.613 ± 2.971  ops/ms
RocksStateBackendBenchmark.stateBackends       ROCKS_INC  thrpt   30  288.627 ± 1.598  ops/ms
```

As we can see, there is a slight impact (1%~2%) in pure keyby performance, but it is not obvious when involving state backends. I think it's acceptable, WDYT?

And I'd rather not checking if the new KeyGroupIndex is the same as the current one, since the overhead is not significantly lower than current solution. WDYT?

Reply:
BTW, apart from changing the base class, we can also only change the `RocksDBKeyedStateBackend#setCurrentKey` to avoid the impact on HashMapStateBackend, as the HashMapStateBackend will always check the key group when accessing state tables.

Reply:
Actually I don't like the idea of checking if current key group is valid when accessing the state, since the problem is from setting the key group instead of accessing the state. Besides, user may set key group once and access the state several times, so for performance concern, I'd rather remove the check in each state access and only keep the check in this PR. WDYT? @Myasuka

Reply:
I'm afraid that we cannot avoid the check on accessing state as users might provide a non-deterministic hashCode for the current key.

Reply:
> I'm afraid that we cannot avoid the check on accessing state as users might provide a non-deterministic hashCode for the current key.

Hi @Myasuka , I checked the code in ```StateTable```, and it seems that in most state accessing cases (expect queryable state) we are checking the key group from ```keyContext.getCurrentKeyGroupIndex()``` instead of calculating it by ```hashCode``` of the partitioned key. So actually we are checking the same value when ```setCurrentKeyGroupIndex``` or state accessing, whether the hashCode implementation is deterministic or not.

Reply:
Maybe we can take a look at FLINK-23908, from which Flink community introduced the key group check.

Reply:
> BTW, apart from changing the base class, we can also only change the RocksDBKeyedStateBackend#setCurrentKey to avoid the impact on HashMapStateBackend, as the HashMapStateBackend will always check the key group when accessing state tables.

How about the opposite: removing the check from HashMapStateBackend (StateTable) and relying on `InternalKeyContext`? So that it works for any backend.
As a side benefit, that would be a bit faster because the check would be done once per input record, not per state access.

Reply:
> How about the opposite: removing the check from HashMapStateBackend (StateTable) and relying on `InternalKeyContext`? So that it works for any backend. As a side benefit, that would be a bit faster because the check would be done once per input record, not per state access.

I Agree. That's what I'm trying to say.

Reply:
I think Roman's suggestion is fine for developers, and for general users, as FLINK-23908 talked about, the current check would be fine during snapshotting. Maybe we can keep this implementation as current Zakelly did in this PR considering the performance impact is so small.

Reply:
Since we agree to keep this change as it is, would you please approve and merge this PR? Or do you have any other comments? Thanks @rkhachatryan @Myasuka .